test/e2e/capi_test.go

```diff
@@ -226,15 +226,30 @@ var _ = Describe("Running the Cluster API E2E tests", func() {
 				})
 				capi_e2e.ClusterctlUpgradeSpec(ctx, func() capi_e2e.ClusterctlUpgradeSpecInput {
 					return capi_e2e.ClusterctlUpgradeSpecInput{
-						E2EConfig:             e2eConfig,
-						ClusterctlConfigPath:  clusterctlConfigPath,
-						BootstrapClusterProxy: bootstrapClusterProxy,
-						ArtifactFolder:        artifactFolder,
-						SkipCleanup:           skipCleanup,
-						PreInit:               getPreInitFunc(ctx),
+						E2EConfig:                 e2eConfig,
+						ClusterctlConfigPath:      clusterctlConfigPath,
+						BootstrapClusterProxy:     bootstrapClusterProxy,
+						ArtifactFolder:            artifactFolder,
+						SkipCleanup:               skipCleanup,
+						PreInit:                   getPreInitFunc(ctx),
+						InitWithProvidersContract: "v1beta1",
 						ControlPlaneWaiters: clusterctl.ControlPlaneWaiters{
 							WaitForControlPlaneInitialized: EnsureControlPlaneInitialized,
 						},
+						InitWithBinary: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.0.5/clusterctl-{OS}-{ARCH}",
+						// We have to pin the providers because with `InitWithProvidersContract` the test would
+						// use the latest version for the contract (which is v1.3.X for v1beta1).
+						InitWithCoreProvider:            "cluster-api:v1.0.5",
+						InitWithBootstrapProviders:      []string{"kubeadm:v1.0.5"},
+						InitWithControlPlaneProviders:   []string{"kubeadm:v1.0.5"},
+						InitWithInfrastructureProviders: []string{"azure:v1.0.2"},
+						// We have to set this to an empty array as clusterctl v1.0 doesn't support
+						// runtime extension providers. If we don't do this the test will automatically
+						// try to deploy the latest version of our test-extension from docker.yaml.
+						InitWithRuntimeExtensionProviders: []string{},
+						InitWithKubernetesVersion:         "v1.23.13",
```

Comment:
why are we hardcoding a k8s version here?

Reply:
I followed the pattern from CAPI's call to the clusterctl upgrade test. Before, we used the `INIT_WITH_KUBERNETES_VERSION` environment variable and I agree it's better to use that. However, I do intend on adding more upgrade tests from different versions of `v1beta1`, and CAPI uses different k8s version for that. I'll default to using the environment variable for now but we may have to hardcode k8s versions for the future upgrade tests:

https://github.com/kubernetes-sigs/cluster-api/blob/main/test/e2e/clusterctl_upgrade_test.go#L110